File: UDP Interface Python/UDP Interface.py
# -*- coding: utf-8 -*-
"""
Created on Sun Jun  6 20:21:33 2021

@author: Chen Mengnan
"""
import UDP_class
import sim_parameters_class
import numpy as np
IP="127.0.0.1"
writePort=8847
readPort=8848
Labview_Receive_Address=(IP,writePort)
Labview_Send_Address=(IP,readPort)

sim=sim_parameters_class.sim_parameters()
u=UDP_class.UDP_class(IP,writePort,readPort)      
u.do_open_UDP()
u.set_TX_parameters(sim,Labview_Receive_Address)
u.set_RX_parameters(sim,Labview_Receive_Address)
for i in range(100):
    print('This is the',i,'th test:')
#   Data generation
#   sim.data=np.random.randint(5,size=sim.data_size) #generate random int with data_size 
    sim.array=np.random.rand(sim.K,sim.M)+1j*np.random.rand(sim.K,sim.M) #generate random complex array
#    sim.array=np.random.rand(sim.K,sim.M)       #generate random real array
    sim.UDP_data_packet_size=np.random.randint(99)
    u.do_UDP_write_data(sim,Labview_Receive_Address)
    u.do_UDP_read_data(sim,Labview_Send_Address)
    u.do_evaluate(sim,Labview_Receive_Address)
    if u.retransmission==1:
        break
    print('-------------------------------------------');
    print(' ');
    print(' ');
    print('' );
u.do_close_UDP()

File: UDP Interface Matlab/UDP_Interface.m
clc;
clear;
UDP.IP='127.0.0.1';
UDP.write_port=8847;
UDP.read_port=8848;
TX_Frequency=2320; %Hz
TX_Sample_Rate=3;
RX_Frequency=2000;
RX_Sample_Rate=2;
sim_parameters.K =64;
sim_parameters.M =4;   
sim_parameters.J = 4;
sim_parameters.waveform.name = 'SWH'; 
sim_parameters.N_iterations = 4;
sim_parameters.waveform.Q = 16;
sim_parameters.waveform.Q2 = 4;
sim_parameters.R = 1/2; 
sim_parameters.punc = [1 1 0 1 1 0];
sim_parameters.encoder = 'PC_convolutional';
SNR = 7; 
sim_parameters.P = 10^(SNR/10);
sim_parameters.data_type=1;  %0: random modulated signal, 1:random bits 0 or 1
sim_parameters.data_size=299; %Number of bits
sim_parameters.equalizer = 'CWCU_MMSE_low_complexity';
sim_parameters.UDP_data_packet_size=60;

%step 1: initalization UDP 
obj = UDP_func(UDP);
%step 2: open UDP
obj = do_open_UDP(obj);
%step 3: set TX parameters and waveform parameters
obj = set_TX_parameters(obj,TX_Frequency,TX_Sample_Rate,sim_parameters);
%step 4: set RX parameters
obj = set_RX_parameters(obj,RX_Frequency,RX_Sample_Rate,sim_parameters);
%data_type=0==> random modulated signal
%data_type=1==> random bits 
Nevents=200;
for j=1:Nevents
disp(['Test process : ',num2str((j/Nevents)*100),'%']);
disp(['This is the ', num2str(j), 'th iteration']);   
%% Data generation
% sim_parameters.matrix = rand(sim_parameters.K,sim_parameters.M)+1i*rand(sim_parameters.K,sim_parameters.M);
sim_parameters.random_bit=randi(2,1,sim_parameters.data_size)-1;
%%
obj = do_UDP_write_data(obj,sim_parameters);
obj = do_UDP_read(obj);
obj = do_evaluate(obj,sim_parameters);
if(obj.retransmission==1)
    break
end
disp('-------------------------------------------');
disp(' ');
disp(' ');
disp('' );
end
obj = do_UDP_close(obj);



















